python-checks/src/main/java/org/sonar/python/checks/hotspots/GraphQLIntrospectionCheck.java

```diff
@@ -0,0 +1,133 @@
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.ExpressionList;
+import org.sonar.plugins.python.api.tree.ListLiteral;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tuple;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6786")
+public class GraphQLIntrospectionCheck extends PythonSubscriptionCheck {
+
+  private static final Set<String> GRAPHQL_VIEWS_FQNS = Set.of(
+    "flask_graphql.GraphQLView.as_view",
+    "graphql_server.flask.GraphQLView.as_view");
+
+  private static final Set<String> SENSITIVE_VALIDATION_RULE_FQNS = Set.of(
+    "graphene.validation.DisableIntrospection",
+    "graphql.validation.NoSchemaIntrospectionCustomRule");
+
+  private static final String MESSAGE = "Disable introspection on this GraphQL server endpoint.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, GraphQLIntrospectionCheck::checkGraphQLIntrospection);
+  }
+
+  private static void checkGraphQLIntrospection(SubscriptionContext ctx) {
+    CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+    Optional.ofNullable(callExpression.calleeSymbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter(fqn -> GRAPHQL_VIEWS_FQNS.contains(fqn) &&
+        (hasSensitiveMiddlewares(callExpression.arguments()) || hasSensitiveValidationRules(callExpression.arguments())))
+      .ifPresent(fqn -> ctx.addIssue(callExpression, MESSAGE));
+  }
+
+  private static boolean hasSensitiveMiddlewares(List<Argument> arguments) {
+    RegularArgument argument = TreeUtils.argumentByKeyword("middleware", arguments);
+    if (argument == null) {
+      return true;
+    }
+    return extractArgumentValues(argument)
+      .map(values -> values.isEmpty() || expressionsNameContainIntrospection(values))
+      .orElse(false);
+  }
+
+  private static boolean hasSensitiveValidationRules(List<Argument> arguments) {
+    RegularArgument argument = TreeUtils.argumentByKeyword("validation_rules", arguments);
+    if (argument == null) {
+      return true;
+    }
+    return extractArgumentValues(argument)
+      .map(values -> values.isEmpty() ||
+        expressionsNameContainIntrospection(values) ||
+        expressionsContainsSensitiveRuleFQN(values))
+      .orElse(false);
+  }
+
+  private static Optional<List<Expression>> extractArgumentValues(RegularArgument argument) {
+    return Optional.of(argument)
+      .map(RegularArgument::expression)
+      .flatMap(GraphQLIntrospectionCheck::expressionsFromListOrTuple);
+  }
+
+  private static Optional<List<Expression>> expressionsFromListOrTuple(Expression expression) {
+    return TreeUtils.toOptionalInstanceOfMapper(ListLiteral.class)
```

Comment:
the `TreeUtils. toOptionalInstanceOf` method could be used there

Reply:
Yes good call thanks!